### ISSUE TITLE
[RFC] Beef IDE class file template support

### ISSUE THREAD

**Issue by hunterbridges** [beefytech/Beef#1224]
**Overview**

Many IDEs have some degree of text templating support for making new files, and allow this to be user-extensible. This is a proposal for adding such a feature into the Beef IDE.

**Proposal**

1. Allow each project or workspace to have a "Templates" folder that is excluded from compilation. This could be implicit, or created by the user
2. Users can create new Beef class files in the Templates folder, which can be used as Templates
3. In Beef IDE, wherever you can make a New Class, an additional menu is added to make New Class From Template
4. Whenever that menu is expanded, all of the classes in the Templates folder are listed
5. When a template is selected from the New Class From Template menu, the user is prompted as normal for the desired new class name
6. The new file is created, and any instances of the default name in the template are replaced with the user's inputted class name

**Comment by bfiete** [comment]
I was looking at T4 templates and I had an idea for implementing this using comptime, so you could basically create generator classes... that could be much more powerful that simple templates.

**Comment by hunterbridges** [comment]
Interesting. Would that allow you to basically create classes which write out a file? The only thing I would worry about is how easy it will be to handle lots of multi line string content. Or are you thinking more like T4 where the document is essentially a text file that allows you to add in control flow/meta code?

**Comment by bfiete** [comment]
Also you mentioned multiline string content- if you use verbatim un-interpolated multiline strings then you don't have to slash any quotes or braces..
ie:
```C#
return new .(
@"""
namespace NAMESPACE
{
	class CLASSNAME
	{
		String str = "";
	}
}
""")..Replace("CLASSNAME", className)..Replace("NAMESPACE", namespaceName);
```

**Comment by hunterbridges** [comment]
oh awesome! Yeah I think what you have in mind would work quite well then!

**Comment by bfiete** [comment]
This is now implemented at 73099e4a04f5674bbf46ff57d0acdb7d9f70cd17
Refer to `System.Compiler.NewClassGenerator` to see how it works. Side effects like file IO and process execution is not yet supported.
